File: src/utxo_manager.cpp
#ifndef UTXO_MANAGER_CPP
#define UTXO_MANAGER_CPP
#include <map>
#include <iostream>
#include <string>
#include "transaction.cpp"
using namespace std;

struct UTXO {
    double amount;
    string owner;
};

class UTXOManager {
public:
    map<pair<string, int>, UTXO> utxo_set;

    void add_utxo(string tx_id, int index, double amount, string owner) {
        utxo_set[{tx_id, index}] = {amount, owner};
    }

    void remove_utxo(string tx_id, int index) {
        utxo_set.erase({tx_id, index});
    }

    double get_balance(string owner) {
        double balance = 0;
        for (auto const& it : utxo_set) {
            auto key = it.first;
            auto utxo = it.second;
            if (utxo.owner == owner) balance += utxo.amount;
        }
        return balance;
    }

    bool exists(string tx_id, int index) {
        return utxo_set.count({tx_id, index}) > 0;
    }
};

#endif

File: src/utxo_manager.py
# src/utxo_manager.py

class UTXOManager:
    def __init__(self):
        # Dictionary mapping (tx_id, index) -> {"amount": float, "owner": str}
        self.utxo_set = {}

    def add_utxo(self, tx_id, index, amount, owner):
        """Add a new UTXO to the set."""
        key = (tx_id, index)
        self.utxo_set[key] = {"amount": amount, "owner": owner}

    def remove_utxo(self, tx_id, index):
        """Remove a UTXO (when spent)."""
        key = (tx_id, index)
        if key in self.utxo_set:
            del self.utxo_set[key]

    def get_balance(self, owner):
        """Calculate total balance for an address."""
        balance = 0.0
        for utxo_data in self.utxo_set.values():
            if utxo_data["owner"] == owner:
                balance += utxo_data["amount"]
        return balance

    def exists(self, tx_id, index):
        """Check if UTXO exists and is unspent."""
        return (tx_id, index) in self.utxo_set

    def get_utxo(self, tx_id, index):
        """Return the UTXO data if it exists."""
        return self.utxo_set.get((tx_id, index))
